### ISSUE TITLE
Year-old data removed from table Catalog Record Bulkloader

### ISSUE THREAD

**Issue by dustymc** [ArctosDB/arctos#8148]
4764 year-old records have been removed from Catalog Record Bulkloader (table: bulkloader).

Data:

[bulkloader_2024_10_02.zip](https://github.com/user-attachments/files/17231724/bulkloader_2024_10_02.zip)

Usernames:

* cconlon (1 records)
* allooman (93 records)
* geneseo2012 (57 records)
* mosack (18 records)
* mrichard3 (1 records)
* ripesnipe (3 records)
* dlm (12 records)
* danienics1989 (2 records)
* tmuller (1 records)
* acorral9 (42 records)
* aolivas (16 records)
* hdorquiz (1 records)
* taxonamie (59 records)
* msaqib1 (50 records)
* cpurcell2 (1 records)
* trsapp (2 records)
* Muriel07 (31 records)
* mgmccutcheon (2 records)
* edlindley (61 records)
* ebraker (1 records)
* jebrad (1 records)
* aerochild (513 records)
* apoitras (49 records)
* jturcotte (23 records)
* argastaldi (67 records)
* bbjohnson3 (118 records)
* samerriman (1 records)
* phovingh (3 records)
* admangadu (33 records)
* kderieg (4 records)
* pauldougherty (47 records)
* brandonsthompson (1 records)
* YiqiXu (22 records)
* kirk0418 (43 records)
* vmata (2 records)
* jegelewicz (4 records)
* jrnapier (1 records)
* cherrera (6 records)
* gmjohnson7 (2 records)
* csluka (13 records)
* dwise (16 records)
* carlslieb (86 records)
* jmaley1 (1 records)
* apizarro (2 records)
* mhilke (6 records)
* emhernandez12 (31 records)
* ahodge (41 records)
* rmatthews (407 records)
* ewommack (26 records)
* mlbrady (1 records)
* sadaniels (22 records)
* sarahmdv (603 records)
* MWMartins (5 records)
* cfilipek (1 records)
* anhoover (56 records)
* claypollock (1721 records)
* sara (38 records)
* erikajprice (235 records)
* nickpineschi (2 records)
* tmpoirrier (57 records)

Users have access to these collections:

* ACUNHC:Bird
* ACUNHC:Ento
* ACUNHC:Fish
* ACUNHC:Herp
* ACUNHC:Inv
* ACUNHC:Mamm
* ACUNHC:Paleo
* ALMNH:Anc
* ALMNH:Arc
* ALMNH:Bird
* ALMNH:EH
* ALMNH:Ento
* ALMNH:ES
* ALMNH:Geo
* ALMNH:Herp
* ALMNH:Inv
* ALMNH:Mamm
* ALMNH:Paleo
* ANSP:Host
* ANSP:Para
* APSU:Bird
* APSU:Fish
* APSU:Herp
* APSU:Mamm
* Arctos:Entity
* ASNHC:Bird
* ASNHC:Herb
* ASNHC:Herp
* ASNHC:Mamm
* ASNHC:Para
* ASUMZ:Bird
* ASUMZ:Bivalve
* ASUMZ:Ento
* ASUMZ:Fish
* ASUMZ:Herp
* ASUMZ:Invert
* ASUMZ:Mamm
* BIOAPT:LEGECC
* BIOAPT:UABB
* BSUNH:Bird
* BSUNH:Fish
* BSUNH:Herb
* BSUNH:Herp
* BSUNH:Inv
* BSUNH:Mamm
* BYU:Bird
* BYU:Edu
* BYU:Fish
* BYU:Herp
* BYU:Mamm
* BYUObs:Bird
* BYUObs:Herp
* BYUObs:Mamm
* BYU:Teach
* CDSC:Arch
* CDSC:Bird
* CDSC:Ento
* CDSC:Geol
* CDSC:Herb
* CDSC:Herp
* CDSC:Inv
* CDSC:Mamm
* CHAS:Art
* CHAS:AV
* CHAS:Bird
* CHAS:Egg
* CHAS:EH
* CHAS:Ento
* CHAS:ES
* CHAS:Fish
* CHAS:Herb
* CHAS:Herp
* CHAS:Inv
* CHAS:Mamm
* CHAS:Teach
* COA:Bird
* COA:Egg
* COA:Ento
* COA:Herp
* COA:Mamm
* COA:Rept
* CRCM:Bird
* CRCM:Fish
* CRCM:Herp
* CRCM:Mamm
* CRCM:Paleo
* CRCM:Teach
* CSULB:Bird
* CSULB:Fish
* CSULB:Herp
* CSULB:Mamm
* DGR:Bird
* DGR:Ento
* DGR:Mamm
* DMNS:Bird
* DMNS:Egg
* DMNS:Herp
* DMNS:Inv
* DMNS:Mamm
* DMNS:Para
* HMCM:Arthro
* HMCM:Bird
* HMCM:Geo
* HMCM:Herb
* HMCM:Herp
* HMCM:Mamm
* HSUVM:Amph
* HSUVM:Mamm
* HSUVM:Rept
* HWML:Para
* JFBM:Fish
* JFBM:Herp
* JFBM:Inv
* JSNM:Egg
* JSNM:Herb
* JSNM:Paleo
* KNWR:Ento
* KNWR:Env
* KNWR:Herb
* KNWR:Inv
* KNWRObs:Bird
* KNWRObs:Fish
* KNWRObs:Herb
* KNWRObs:Mamm
* KSB:Mamm
* KSB:Teach
* KWP:Ento
* LINGU:Herp
* MLZ:Bird
* MLZ:Egg
* MLZ:Fish
* MLZ:Herb
* MLZ:Inv
* MLZ:Mamm
* MLZObs:Fish
* MMNH:Bird
* MMNH:Edu
* MMNH:Mamm
* MMNS:Birds
* MMNS:CO
* MMNS:Crust
* MMNS:Ento
* MMNS:Fishes
* MMNS:GRC
* MMNS:Herp
* MMNS:IP
* MMNS:IZ
* MMNS:Mala
* MMNS:Mamm
* MMNS:PB
* MMNS:Photo
* MMNS:RM
* MMNS:VP
* MSB:Arth
* MSB:Bird
* MSB:DGR
* MSB:Fish
* MSB:Herb
* MSB:Herp
* MSB:Host
* MSB:Inv
* MSB:Mamm
* MSBObs:Herp
* MSBObs:Mamm
* MSB:Para
* MVZ:Arch
* MVZ:Bird
* MVZ:Egg
* MVZ:Fish
* MVZ:Herp
* MVZ:Hild
* MVZ:Img
* MVZ:Mamm
* MVZObs:Bird
* MVZObs:Fish
* MVZObs:Herp
* MVZObs:Mamm
* NBSB:Bird
* NCWRC:Bird
* NCWRC:Bivalve
* NCWRC:Crust
* NCWRC:Ento
* NCWRC:Fish
* NCWRC:Gastro
* NCWRC:Herp
* NCWRC:Mamm
* NHSM:Arc
* NHSM:Bird
* NHSM:Ento
* NHSM:Geol
* NHSM:Herb
* NHSM:Herp
* NHSM:Inv
* NHSM:Mamm
* NHSM:Paleo
* NMMNH:Bird
* NMMNH:Ento
* NMMNH:Fish
* NMMNH:Geol
* NMMNH:Herb
* NMMNH:Herp
* NMMNH:Inv
* NMMNH:Mamm
* NMMNH:Obs
* NMMNH:Paleo
* NMU:Bird
* NMU:Ento
* NMU:Fish
* NMU:Herb
* NMU:Herp
* NMU:Mamm
* NMU:Para
* OGL:Genomic
* OWU:Amph
* OWU:Bird
* OWU:EH
* OWU:Ento
* OWU:ES
* OWU:Fish
* OWU:Geol
* OWU:Herb
* OWU:Inv
* OWU:Mamm
* OWU:Para
* OWU:Rept
* PSM:Alg
* PSM:Arch
* PSM:Bird
* PSM:Ento
* PSM:Geol
* PSM:Herb
* PSM:Herp
* PSM:Inv
* PSM:Mamm
* PSM:Paleo
* PSM:Teach
* PSU:Mamm
* RWLC:Bird
* RWLC:Fish
* RWLC:Herp
* RWLC:Mamm
* STAR:Alg
* STAR:Herb
* TCDGM:Mineral
* TCDGM:Paleo
* TCDGM:Rock
* TMCC:Bird
* TMCC:Ento
* TMCC:Fish
* TMCC:Herb
* TMCC:Herp
* TMCC:Inv
* TMCC:Mamm
* TMCC:Teach
* UAA:Ento
* UAF:Zooarch
* UAM:Alg
* UAM:Arc
* UAM:Art
* UAMb:Herb
* UAM:Bird
* UAM:EH
* UAM:Ento
* UAM:Env
* UAM:ES
* UAM:Fish
* UAM:Herb
* UAM:HerbLib
* UAM:Herp
* UAM:Inv
* UAM:Mamm
* UAM:Myco
* UAMObs:Art
* UAMObs:Bird
* UAMObs:EH
* UAMObs:Ento
* UAMObs:Fish
* UAMObs:Mamm
* UA:Paleo
* UCM:Bird
* UCM:Egg
* UCM:Ento
* UCM:Fish
* UCM:FossilEgg
* UCM:FossilTrack
* UCM:FossilVert
* UCM:Herp
* UCM:Inv
* UCM:IP
* UCM:Mamm
* UCM:Obs
* UCM:Para
* UCM:PB
* UCONN:Bird
* UCONN:Egg
* UCSC:Bird
* UCSC:Fish
* UCSC:Herp
* UCSC:Mamm
* UMNH:Bird
* UMNH:Ento
* UMNH:Herp
* UMNH:Host
* UMNH:Mala
* UMNH:Mamm
* UMNHObs:Mamm
* UMNH:Para
* UMNH:Teach
* UMZM:Bird
* UMZM:Egg
* UMZM:Mamm
* UMZM:Para
* UNCG:Mamm
* UNCG:Para
* UNM:Geol
* UNM:MET
* UNM:Paleo
* UNR:Bird
* UNR:Egg
* UNR:Fish
* UNR:Herp
* UNR:Mamm
* UNR:Teach
* USNPC:Para
* UTEP:Arc
* UTEP:Bird
* UTEP:Ento
* UTEP:ES
* UTEP:Fish
* UTEP:Herb
* UTEP:Herp
* UTEP:HerpOS
* UTEP:Inv
* UTEP:Mamm
* UTEPObs:Bird
* UTEPObs:Ento
* UTEPObs:Herb
* UTEPObs:Herp
* UTEP:Teach
* UTEP:Zoo
* UWBM:Herp
* UWBM:IP
* UWBM:Mamm
* UWBM:MET
* UWBM:MINERAL
* UWBM:PaleoArchive
* UWBM:PB
* UWBM:PR
* UWBM:VP
* UWYMV:Bird
* UWYMV:Egg
* UWYMV:Fish
* UWYMV:Herp
* UWYMV:Mamm
* UWZM:Bird
* UWZM:EH
* UWZM:Fish
* UWZM:Herp
* UWZM:Inv
* UWZM:Mamm
* UWZM:Osteo
* UWZM:Paleo
* UWZM:Teach
* WNMU:Bird
* WNMU:Fish
* WNMU:Mamm
* ZSFQ:Herp

Collection Contacts

* @ebraker
* @bryansmclean
* @ehalverson26
* @Nicole-Ridgwell-NMMNHS
* @mkoo
* @ronaldeng
* @AJLinn
* @campmlc
* @ccicero
* @WaigePilson
* @camwebb
* @amgunderson
* @atrox10
* @DerekSikes
* @mvzhuang
* @ufarrell
* @cjconroy
* @rwilhoyt
* @jtgiermakowski
* @wellerjes
* @asemerdj
* @keg34
* @RJSS8
* @justybob
* @jebrad
* @brandon-s-thompson
* @babogan
* @gradyjt
* @jrpletch
* @AdrienneRaniszewski
* @aklompma
* @acdoll
* @jldunnum
* @jrdemboski
* @happiah-madson
* @ccwlobo
* @byuherpetology
* @genevieve-anderegg
* @msbparasites
* @mlbowser
* @leet1984
* @javanveldhuizen
* @jessicatir
* @DellaCHall
* @falco-rk
* @megulbranson
* @ewommack
* @kderieg322079
* @sharpphyl
* @StefanieBond
* @catherpes
* @MTArmour
* @gracz-UNL
* @kat-sterner
* @sjshirar
* @lin-fred
* @KatherineLAnderson
* @adhornsby
* @kmkocot
* @jandreslopez
* @droberts49
* @shorrack

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
Why??

**Comment by dustymc** [comment]
>Why??

https://github.com/ArctosDB/arctos/issues/6398

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
Can the NMMNH:Paleo records please be reloaded? These are all data that I am working through and will get to.

**Comment by jebrad** [comment]
Thanks Dusty.  For my clarification: were these removed data put anyplace else besides this morning's "bulkloader_2024_10_02.zip"?   
I assume No, but your comment from https://github.com/ArctosDB/arctos/issues/6398 makes me wonder: 
"Nothing that happens here will result in any data going away, and I apologize if that wasn't clear. I'd only like to move some maybe-abandoned data somewhere that doesn't interfere with my work."

I have all of mine from this morning's zip file, and I will work through and get them re-entered or bulkloaded.   Anything that was entered over a year ago was probably done wrong, or not following our new procedures, so this makes sense for me to do.

(Also I will try to be better about cleaning out my data in the bulkloader, I rarely consider its effect on your work.)

**Comment by dustymc** [comment]
@Nicole-Ridgwell-NMMNHS they should re-load, are you asking me to do that? Is there some better format-or-whatever for next time? This isn't meant to be disruptive, but neither can the loaders be everyone's (insecure) ministorage, I'm very open for ideas regarding how any of this might be made better. 

6 month warnings and year deletions??

Monthly year-old purges instead of yearly???

@jebrad correct, the data are here in the zip file, no longer in the table.

>over a year ago was probably done wrong

Yea, that's kinda what lead here, most (clearly not all, surprise!) things over some age seem to permanently fall off the radar.

**Comment by campmlc** [comment]
Many of the records for MSB were stranded there when changes were made to the bulkloader and to controlled vocabulary and to permissions.  It helps to keep them in the bulkloader to keep them on the radar, as many of them require significant time and dedicated attention to fix, both items in short suppy. It might be helpful going forward if we could request the possibility of database updates including the bulkloader as well as existing records?

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
My issue is that we are currently at almost a two year delay between cataloging and final data upload (which represents almost 10,000 records in limbo). I HATE this state of affairs and getting records cleaned and loaded is my top priority, but it doesn't seem to be anyone else's priority so I get loaded up with a ton of other stuff to do, usually "please catalog more stuff". Most of this data is in excel files that came directly from the researchers that I clean and load. This goes to the bulkloader and gets reviewed and loaded to the database immediately. About 10-15% is data I have undergrads and volunteers entering directly from specimens/tags as the specimens are being cataloged. That goes into the bulkloader and waits for it's spot in my queue. I have a master spreadsheet to keep track of where the data is for un-databased records. I am slowly making a dent in this gap, but it will be awhile before I have it under one year.

Basically I'm just annoyed to have to take the time to have to reupload into the bulkloader (especially making sure dates don't get mangled in the process) or just keep the data dump locally, but have to update a bunch of entries in my tracking worksheet.

In other words, I'm buried under a mountain trying to dig myself out and grumbling at a small bucket of rocks being dumped in my hole.

**Comment by Jegelewicz** [comment]
I'd like to suggest that instead of just removing stuff, an announcement could go out giving people two weeks to respond. @Nicole-Ridgwell-NMMNHS could have said - I'm still working - please leave my stuff and then she wouldn't have to do the work twice. No response means you are OK with the removal of the data.

An email and a banner would be a good start.

**Comment by leet1984** [comment]
Dear All,

I am still not sure why data needs to be removed in the first place.

Sincerely,
Tom Lee

**Comment by campmlc** [comment]
I agree that the process of re-bulkloading from a downloaded csv is extremely onerous and difficult, because we have to wade through hundreds of empty data columns and fix hundreds of corrupted dates. Not to mention the possibility of introducing the dreaded invisible non-printing characters in the process. Currently when the data are in browse and edit, we can select to remove empty columns from display. If there were a way to remove them from download as well, that would be a huge step forward. If we could donwload such that the date formats could be preserved, that would be immensely valuable. Then we could download data to get it out of the way and work on it locally, make corrections to updated controlled vocabulary etc, and then reupload it easily.

**Comment by dustymc** [comment]
>fix hundreds of corrupted dates.

If you've got that it's because your tools are doing it. (If that's not the case then PLEASE tell me how to re-create and I'll put that on the very top of my bug pile.)

>way to remove them from download

That would also be up to your local tools. Here's a how-to for one easily accessible tool: https://developers.google.com/workspace/add-ons/samples/clean-sheet

**Comment by dustymc** [comment]
See also https://handbook.arctosdb.org/documentation/dates.html#excel-issues

**Comment by campmlc** [comment]
I think I speak for a majority of users who have to manage these files in Excel. I don't have the resources and mental space to learn new software on top of trying to manage the daily job onslaught as @Nicole-Ridgwell-NMMNHS describes - this is why we use Arctos. Lack of time and resources to address these records is how they came to be left in the bulkloader. If Arctos could provide some tools to make it easier for users to manage these data, that would be very helpful. In the absence of that, I also agree that users should be given the option of leaving data in the bulkloader.

**Comment by campmlc** [comment]
Perhaps Arctos could integrate some of the google tools referenced here?

**Comment by WaigePilson** [comment]
> If you've got that it's because your tools are doing it. (If that's not the case then PLEASE tell me how to re-create and I'll put that on the very top of my bug pile.)

This is a bit of an aside from this core issue, but when I format my dates to YYYY-MM-DD, save as csv UTF-8, and exit the spreadsheet, when I reopen that csv in excel it will change all the dates back to M/D/YYYY. I can re-format the cells again, save, and as long as I don't open that file again it will leave the dates as YYYY-MM-DD and Arctos will accept the upload. But if I open the file in excel, dates get reset. So (in this case) if you download data OUT of Arctos and open the csv, the dates will reset and you'll have to re-format them again (this could be across 10+ columns randomly spread across your 200+ columns of data if you're uploading specimen records). If there is a way to save a csv and have it keep the fidelity of your date formatting when you re-open it, please enlighten!

**Comment by DerekSikes** [comment]
I think if you save the file as excel format (.xlsx) rather than .csv when you open it again it will have kept your YYYY-MM-DD formatting intact. This might be a way to avoid reformatting - keep a master in excel format and only 'export' to csv for upload to Arctos.

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
Usually what works for me is to to open a blank excel sheet, go to the data tab and select import as csv, click transform data, select all columns and change data type to text.

**Comment by Nicole-Ridgwell-NMMNHS** [comment]
> neither can the loaders be everyone's (insecure) ministorage

For folks that need to keep their data in the bulkloader, the yearly opt-out of the data removal could also serve as a reminder to download a backup copy of any data they have in there.

**Comment by mkoo** [comment]
HI all, weighing in now:

>>Why?

The main reasons are all technical -- here are the big ones:
- It's a (potential) resource hog; records sitting in the loader accumulates and makes everything slower for everyone. This round was ca. 5k records; last time was about 20k (after a few years) so you can imagine that it can balloon in unpredictable ways. If kept to a small potbelly pig size, it's fine, but we are trying to avoid a Hogzilla 
- There's No advantage to having records in bulkloader– not viewable, not auto-fixable (ie. data doesnt get fixed as it marinates in the loader). I understand that there's format cleaning and other data issues going on and reasons why records sit there, which is a great opportunity for some new enhancements or help-- may be send data to google spreadsheets far away from the Excel and its evil date formatting... So let's brainstorm new tools and best practices separate from this issue.
- Lastly, data in the bulkloader are NOT backed-up. It is meant to be a temporary holding table for curatorial review prior to record creation. An ephemeral resource at best for checking and convenient data fixes.

I think Dusty said that some notifications can be set up for users so there are reminders instead of issues filed which can be worked out. I propose a regular yearly schedule for clearing out data in the bulkloader that has reached its first birthday (easier on programmer and predictable for users). Open for feedback on the schedule but yearly clearing is necessary to keep this under control.

**Comment by Jegelewicz** [comment]
Note to anyone with this data - opening the file in Numbers or Excel will result in a truncated file as they can only handle 1000 columns. Use Google Sheets and remove empty columns to get below 1000 before using Excel or Numbers.
